meminfo.go

```diff
@@ -0,0 +1,174 @@
+	WritebackTmp int64 `meminfo:"WritebackTmp"`
+	// Based on the overcommit ratio ('vm.overcommit_ratio'),
+	// this is the total amount of  memory currently available to
+	// be allocated on the system. This limit is only adhered to
+	// if strict overcommit accounting is enabled (mode 2 in
+	// 'vm.overcommit_memory').
+	// The CommitLimit is calculated with the following formula:
+	// CommitLimit = ([total RAM pages] - [total huge TLB pages]) *
+	//                overcommit_ratio / 100 + [total swap pages]
+	// For example, on a system with 1G of physical RAM and 7G
+	// of swap with a `vm.overcommit_ratio` of 30 it would
+	// yield a CommitLimit of 7.3G.
+	// For more details, see the memory overcommit documentation
+	// in vm/overcommit-accounting.
+	CommitLimit int64 `meminfo:"CommitLimit"`
+	// The amount of memory presently allocated on the system.
+	// The committed memory is a sum of all of the memory which
+	// has been allocated by processes, even if it has not been
+	// "used" by them as of yet. A process which malloc()'s 1G
+	// of memory, but only touches 300M of it will show up as
+	// using 1G. This 1G is memory which has been "committed" to
+	// by the VM and can be used at any time by the allocating
+	// application. With strict overcommit enabled on the system
+	// (mode 2 in 'vm.overcommit_memory'),allocations which would
+	// exceed the CommitLimit (detailed above) will not be permitted.
+	// This is useful if one needs to guarantee that processes will
+	// not fail due to lack of memory once that memory has been
+	// successfully allocated.
+	CommittedAS int64 `meminfo:"Committed_AS"`
+	// total size of vmalloc memory area
+	VmallocTotal int64 `meminfo:"VmallocTotal"`
+	// amount of vmalloc area which is used
+	VmallocUsed int64 `meminfo:"VmallocUsed"`
+	// largest contiguous block of vmalloc area which is free
+	VmallocChunk      int64 `meminfo:"VmallocChunk"`
+	HardwareCorrupted int64 `meminfo:"HardwareCorrupted"`
+	AnonHugePages     int64 `meminfo:"AnonHugePages"`
+	HugePagesTotal    int64 `meminfo:"HugePages_Total"`
+	HugePagesFree     int64 `meminfo:"HugePages_Free"`
+	HugePagesRsvd     int64 `meminfo:"HugePages_Rsvd"`
+	HugePagesSurp     int64 `meminfo:"HugePages_Surp"`
+	Hugepagesize      int64 `meminfo:"Hugepagesize"`
+	DirectMap4k       int64 `meminfo:"DirectMap4k"`
+	DirectMap2M       int64 `meminfo:"DirectMap2M"`
+}
+
+// NewMeminfo returns kernel/system statistics read from /proc/stat.
+func NewMeminfo() (Meminfo, error) {
+	fs, err := NewFS(DefaultMountPoint)
+	if err != nil {
+		return Meminfo{}, err
+	}
+
+	return fs.NewMeminfo()
+}
+
+// NewMeminfo returns an information about current kernel/system statistics.
+func (fs FS) NewMeminfo() (m Meminfo, err error) {
+	f, err := os.Open(fs.Path("meminfo"))
+	if err != nil {
+		return Meminfo{}, err
+	}
+	defer f.Close()
+
+	st := reflect.TypeOf(m)
+
+	s := bufio.NewScanner(f)
+	for s.Scan() {
+		line := s.Text()
+
+		re := regexp.MustCompile(m.Regex())
```

Comment:
It'd be better to not compile a new regex during every invocation. I'd move that into a global variable:

```
var memInfoRE = regexp.MustCompile("....")
```